File: arraysAndHashing/groupAnagrams.py
# based on sorting algo time complexity is O(m*n*logn) and space complexity is O(n)
# hashmap algo time complexity is O(n*m*26) ~ O(n*m)
from collections import defaultdict

class Solution(object):
    def groupAnagrams(self, strs):
        hashMap = defaultdict(list) # list of mapping charCount to List of anagrams
        for s in strs:
            count = [0] * 26 # a - z
            for c in s:
                count[ord(c) - ord('a')] += 1
            hashMap[tuple(count)].append(s)
        return hashMap.values()
        

if __name__ == "__main__":
    strs = ["eat", "tea", "tan", "ate", "nat", "bat"]
    print(Solution().groupAnagrams(strs))

File: arraysAndHashing/cppRealizations/groupAnagrams.cpp
#include <vector>
#include <string>
#include <map>
#include <iostream>
#include <algorithm>
#include <tuple>

using namespace std;

// overload operator << for vector
template <typename T>
ostream& operator<<(ostream& out, const vector<T>& str){
    out << "[";
    for(size_t i = 0; i < str.size(); i++){
        if(i < str.size() - 1){
            out << str[i] << ",";
        }else{
            out << str[i];
        }
    }
    out << "]";
    return out;
}

vector<vector<string>> groupAnagrams(vector<string>& strs) {
    unordered_map<string, vector<string>> hashMap;
    for(const string& s : strs){
        vector<int> count(26, 0);
        for(char c : s){
            count[c - 'a']++;
        }
        // convert vector<int> to string for inserting to the unordered map
        string key = "";
        for(int num : count){
            key += to_string(num) + ",";
        }
        // remove last comma
        key.pop_back();
        // insert to the unordered map
        hashMap[key].push_back(s);
    }
    vector<vector<string>> result;
    for(const auto& item : hashMap){
        result.push_back(item.second);
    }
    return result;
}

int main(){
    vector<string> strs = {"eat","tea","tan","ate","nat","bat"};
    vector<vector<string>> result = groupAnagrams(strs);
    cout << result << endl;
    return 0;
}